Fix SD slot numbering.

eMMC -> mmcblk2
microSD -> mmcblk1

Signed-off-by: Fabien Marteau <fabien.marteau@armadeus.com>

---
Index: linux-3.16/drivers/mmc/card/block.c
===================================================================
--- linux-3.16.orig/drivers/mmc/card/block.c	2014-10-02 12:10:07.741220869 +0200
+++ linux-3.16/drivers/mmc/card/block.c	2014-10-02 14:08:25.843109903 +0200
@@ -2162,7 +2162,7 @@
 	 */
 
 	snprintf(md->disk->disk_name, sizeof(md->disk->disk_name),
-		 "mmcblk%d%s", md->name_idx, subname ? subname : "");
+		 "mmcblk%d%s", card->host->index, subname ? subname : "");
 
 	if (mmc_card_mmc(card))
 		blk_queue_logical_block_size(md->queue.queue,
