Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>
Signed-off-by: Jeremie Scheer <jeremie.scheer@armadeus.com>

Index: linux-2.6.38-rc3/arch/arm/plat-mxc/include/mach/board-apf27.h
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.6.38-rc3/arch/arm/plat-mxc/include/mach/board-apf27.h	2011-03-28 18:07:48.000000000 +0200
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2008-2010 Armadeus Systems
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
+ * MA 02110-1301, USA.
+ */
+
+#ifndef __ASM_ARCH_MXC_BOARD_APF27_H__
+#define __ASM_ARCH_MXC_BOARD_APF27_H__
+
+#include <mach/iomux-mx27.h>
+
+/* mandatory (?) for CONFIG_LL_DEBUG */
+#define MXC_LL_UART_PADDR       UART1_BASE_ADDR
+#define MXC_LL_UART_VADDR       AIPI_IO_ADDRESS(UART1_BASE_ADDR)
+
+/* CS5 */
+#define CS5_BASE_ADDR_VIRT 0xF4300000
+#define CS5_SIZE           SZ_1M
+
+/* FPGA addresses */
+#define APF27_FPGA_VIRT (CS5_BASE_ADDR_VIRT)
+#define APF27_FPGA_PHYS (MX27_CS5_BASE_ADDR)
+#define ARMADEUS_FPGA_BASE_ADDR_VIRT (APF27_FPGA_VIRT)
+#define ARMADEUS_FPGA_BASE_ADDR_PHYS (APF27_FPGA_PHYS)
+
+/* FPGA IRQ */
+#define APF27_FPGA_INT_PIN  (GPIO_PORTF | 12)
+#define APF27_FPGA_IRQ      (IRQ_GPIOF(12))
+#define ARMADEUS_FPGA_IRQ   (APF27_FPGA_IRQ)
+
+#endif /* __ASM_ARCH_MXC_BOARD_APF27_H__ */
