
Adds support for APF27's PPS baseboard.

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: linux-2.6.38.1/arch/arm/mach-imx/pps.c
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.6.38.1/arch/arm/mach-imx/pps.c	2011-06-23 17:39:36.000000000 +0200
@@ -0,0 +1,262 @@
+ /*
+ * pps.c
+ * Support of PPS baseboard
+ *
+ * Copyright (C) 2009-2011 ARMadeus Systems
+ * Authors: Julien Boibessot <julien.boibessot@armadeus.com>
+ *          Nicolas Colombain <nicolas.colombain@armadeus.com>
+ *          Fabien Marteau <fabien.marteau@armadeus.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+#include <linux/platform_device.h>
+
+#include <mach/iomux-mx27.h>
+#include <mach/gpio.h>
+#include <mach/board-apf27.h>
+#include <mach/mxc_ehci.h>
+#include <mach/ulpi.h>
+
+#include "devices-imx27.h"
+
+/* Wi2Wi */
+#define BT_RESETn	(GPIO_PORTE | 11)	/* UART3_RTS */
+#define WIFI_PWRDNn	(GPIO_PORTE | 10)	/* UART3_CTS */
+
+static int bt_wifi_pins[] = {
+	(BT_RESETn | GPIO_OUT | GPIO_GPIO),
+	(WIFI_PWRDNn | GPIO_OUT | GPIO_GPIO),
+};
+
+#define USBH2_RST_n_GPIO	(GPIO_PORTB | 10)
+
+static int pps_pins[] = {
+	/* SDHC2 (WiFi) */
+	PB4_PF_SD2_D0,
+	PB5_PF_SD2_D1,
+	PB6_PF_SD2_D2,
+	PB7_PF_SD2_D3,
+	PB8_PF_SD2_CMD,
+	PB9_PF_SD2_CLK,
+	/* USBH1 */
+	PB22_PF_USBH1_SUSP,
+	PB23_PF_USB_PWR,
+	PB24_PF_USB_OC,
+	PB25_PF_USBH1_RCV,
+	PB26_PF_USBH1_FS,
+	PB27_PF_USBH1_OE,
+	PB28_PF_USBH1_TXDM,
+	PB29_PF_USBH1_TXDP,
+	PB30_PF_USBH1_RXDM,
+	PB31_PF_USBH1_RXDP,
+	/* USBH2 */
+	PA0_PF_USBH2_CLK,
+	PA1_PF_USBH2_DIR,
+	PA3_PF_USBH2_NXT,
+	PA4_PF_USBH2_STP,
+	PD22_AF_USBH2_DATA0,
+	PD24_AF_USBH2_DATA1,
+	PD23_AF_USBH2_DATA2,
+	PD20_AF_USBH2_DATA3,
+	PD19_AF_USBH2_DATA4,
+	PD26_AF_USBH2_DATA5,
+	PD21_AF_USBH2_DATA6,
+	PA2_PF_USBH2_DATA7,
+	(USBH2_RST_n_GPIO | GPIO_GPIO | GPIO_OUT),
+};
+
+static int bt_wifi_init(void)
+{
+	int res=0;
+
+	res = mxc_gpio_setup_multiple_pins(bt_wifi_pins,
+			ARRAY_SIZE(bt_wifi_pins), "BT_WIFI");
+	if (!res) {
+		imx27_add_mxc_mmc(1, NULL);
+		gpio_set_value(WIFI_PWRDNn, 1);	/* WiFi on */
+		gpio_set_value(BT_RESETn, 1);	/* Bluetooth on */
+	}
+
+	/* These reset pins are driven by userspace afterwards */
+	mxc_gpio_release_multiple_pins(bt_wifi_pins, ARRAY_SIZE(bt_wifi_pins));
+
+	return res;
+}
+
+static int pps_usbh1_init(struct platform_device *pdev)
+{
+#if 0
+	uint32_t temp;
+
+	temp = readl(IO_ADDRESS(OTG_BASE_ADDR) + 0x600);
+	pr_debug("USB_CTRL before: 0x%08x\n", temp);
+	temp &= ~( (3 << 13) | (1 << 8) |  1 );
+	temp |=  (1 << 4) | (1 << 11) ;
+	writel(temp, IO_ADDRESS(OTG_BASE_ADDR) + 0x600);
+	pr_debug("USB_CTRL after: 0x%08x\n", temp);
+
+	temp = readl(IO_ADDRESS(OTG_BASE_ADDR) + 0x384);
+	pr_debug("PORTSC1 before: 0x%08x\n", temp);
+	temp &= ~(3 << 30);
+	temp |= 3 << 30;
+	writel(temp, IO_ADDRESS(OTG_BASE_ADDR) + 0x384);
+	pr_debug("PORTSC1 after: 0x%08x\n", temp);
+	mdelay(10);
+
+	temp = readl(IO_ADDRESS(OTG_BASE_ADDR) + 0x340);
+	temp |= 0x02; /* reset controller */
+	writel(temp, IO_ADDRESS(OTG_BASE_ADDR) + 0x340);
+	pr_debug("reset controller: 0x%08x\n", temp);
+	mdelay(10);
+#endif
+	return mx27_initialize_usb_hw(pdev->id, MXC_EHCI_POWER_PINS_ENABLED);
+}
+
+struct mxc_usbh_platform_data usbh1_pdata = {
+	.init = pps_usbh1_init,
+	.portsc = MXC_EHCI_MODE_SERIAL,
+};
+
+/*  USBH2  */
+/* define USB registers XXX */
+#define USB_CTRL	(IO_ADDRESS(OTG_BASE_ADDR) + 0x600)
+#define UOG_PORTSC1	(IO_ADDRESS(OTG_BASE_ADDR) + 0x184)
+#define UOG_USBCMD 	(IO_ADDRESS(OTG_BASE_ADDR) + 0x140)
+#define UOG_ULPIVIEW	(IO_ADDRESS(OTG_BASE_ADDR) + 0x170)
+#define UH2_PORTSC1	(IO_ADDRESS(OTG_BASE_ADDR) + 0x584)
+#define UH2_USBCMD	(IO_ADDRESS(OTG_BASE_ADDR) + 0x540)
+#define UH2_ULPIVIEW	(IO_ADDRESS(OTG_BASE_ADDR) + 0x570)
+
+#if 0
+static int usb3315_set_vbus_power(void __iomem *view, int on)
+{
+	int vid, pid, ret = 0;
+
+	vid = (ulpi_read(USB3315_VID_HIGH, view) << 8) |
+		 ulpi_read(USB3315_VID_LOW, view);
+	pid = (ulpi_read(USB3315_PID_HIGH, view) << 8) |
+		 ulpi_read(USB3315_PID_LOW, view);
+
+	pr_info("ULPI h2 Vendor ID 0x%x    Product ID 0x%x\n", vid, pid);
+	if (vid != 0x424  || pid != 0x6) { //XXX
+		pr_err("No USB3315 found\n");
+		return -1;
+	}
+
+	if (on) {
+		/* force disconnection of the DM/DP signals. This
+			avoid powering the connected device through these lines */
+		ulpi_set(RESET, USB3315_FCNCTL, view);
+		mdelay(10); /*wait until the supply is really down*/
+		ulpi_clear(RESET, USB3315_FCNCTL, view);
+		mdelay(1);
+
+		ret = ulpi_set(DRV_VBUS_EXT |	/* enable external Vbus */
+			    DRV_VBUS |		/* enable internal Vbus */
+			    USE_EXT_VBUS_IND |	/* use external indicator */
+			    CHRG_VBUS,		/* charge Vbus */
+			    USB3315_OTGCTL, view);
+	} else {
+		ret = ulpi_clear(DRV_VBUS_EXT |	/* disable external Vbus */
+			      DRV_VBUS,		/* disable internal Vbus */
+			      USB3315_OTGCTL, view);
+
+		ret |= ulpi_set(USE_EXT_VBUS_IND | /* use external indicator */
+			    DISCHRG_VBUS,	   /* discharge Vbus */
+			    USB3315_OTGCTL, view);
+	}
+
+	return ret;
+}
+
+static void usb3315_suspend(void __iomem *view)
+{
+	/* usb3315_set_vbus_power(view, 0); needed ? */
+	ulpi_clear(SUSPENDM, USB3315_FCNCTL, view);
+}
+#endif
+
+static int pps_usbh2_phy_init(struct platform_device *pdev)
+{
+	gpio_set_value(USBH2_RST_n_GPIO, 1);	/* reset off */
+#if 0
+	uint32_t temp;
+	temp = readl(USB_CTRL);
+
+	pr_debug("USB_CTRL before: 0x%08x\n", temp);
+	/* H2SIC = 3 */
+	temp &= ~( (3 << 21) | 1 );
+	/* H2PM=1, H2WIE=1, H2UIE=1 */
+	temp |= (1 << 16) | (1 << 19) | (1 << 20);
+	writel(temp, USB_CTRL);
+	pr_debug("USB_CTRL after: 0x%08x\n", temp);
+
+	temp = readl(UH2_PORTSC1);
+	pr_debug("PORTSC1 before: 0x%08x\n", temp);
+	/* XXX: ? */
+	temp &= ~(3 << 30);
+	temp |= 2 << 30;
+	writel(temp, UH2_PORTSC1);
+	pr_debug("PORTSC1 after: 0x%08x\n", temp);
+
+	mdelay(10);
+
+	temp = readl(UH2_USBCMD);
+	temp |= 0x02; /* reset controller */
+	writel(temp, UH2_USBCMD);
+	pr_debug("reset controller: 0x%08x\n", temp);
+	mdelay(50);
+
+	ret = usb3315_set_vbus_power((void __iomem *)UH2_ULPIVIEW, 1);
+#endif
+
+	return mx27_initialize_usb_hw(pdev->id, MXC_EHCI_INTERFACE_DIFF_UNI);
+}
+
+static struct mxc_usbh_platform_data usbh2_pdata __initdata = {
+	.init = pps_usbh2_phy_init,
+	.portsc = MXC_EHCI_MODE_ULPI,
+};
+
+void apf27_baseboard_resume(void)
+{
+	/* TO BE DONE */
+}
+
+void apf27_baseboard_suspend(void)
+{
+	/* TO BE DONE */
+}
+
+void __init apf27_baseboard_init(void)
+{
+	int ret;
+
+	printk("\n** Registering PPS ressources: **\n");
+	ret = mxc_gpio_setup_multiple_pins(pps_pins,
+			ARRAY_SIZE(pps_pins), "PPS");
+	if (ret)
+		printk(KERN_ERR "pps: Failed to setup pins (%d)\n", ret);
+
+	/* init USB */
+	imx27_add_mxc_ehci_hs(1, &usbh1_pdata);
+	imx27_add_mxc_ehci_hs(2, &usbh2_pdata);
+
+	/* init Wi2Wi */
+	bt_wifi_init();
+
+	printk("done\n");
+}
+
Index: linux-2.6.38.1/arch/arm/mach-imx/Kconfig
===================================================================
--- linux-2.6.38.1.orig/arch/arm/mach-imx/Kconfig	2011-06-15 12:24:58.000000000 +0200
+++ linux-2.6.38.1/arch/arm/mach-imx/Kconfig	2011-06-23 17:19:57.000000000 +0200
@@ -390,6 +390,15 @@
 	  activates the devices you want to use if you have the "Full"
 	  version.
 
+config MACH_PPS
+	prompt "Armadeus Systems APF27 PPS Secured Gateway"
+	bool
+	depends on MACH_APF27
+	select MACH_APF27_HAS_BASEBOARD
+	select IMX_HAVE_PLATFORM_MXC_MMC
+	help
+	  This selects the APF27_PPS as baseboard.
+
 endchoice
 
 config MACH_APF27_DEV_EXT
Index: linux-2.6.38.1/arch/arm/mach-imx/Makefile
===================================================================
--- linux-2.6.38.1.orig/arch/arm/mach-imx/Makefile	2011-06-15 12:24:58.000000000 +0200
+++ linux-2.6.38.1/arch/arm/mach-imx/Makefile	2011-06-23 17:39:11.000000000 +0200
@@ -42,3 +42,4 @@
 obj-$(CONFIG_MACH_APF27_DEV) += apf27-dev.o
 obj-$(CONFIG_MACH_APF27_DEV) += apf27-lcd.o
 obj-$(CONFIG_MACH_APF27_DEV_EXT_WIRELESS) += apf27-dev-ext-apw.o
+obj-$(CONFIG_MACH_PPS) += pps.o
