Index: linux-2.6.38.8/drivers/Kconfig
===================================================================
--- linux-2.6.38.8.orig/drivers/Kconfig	2011-10-20 18:47:06.000000000 +0200
+++ linux-2.6.38.8/drivers/Kconfig	2011-10-20 18:47:40.000000000 +0200
@@ -117,4 +117,7 @@
 source "drivers/platform/Kconfig"
 
 source "drivers/clk/Kconfig"
+
+source "drivers/mxc/Kconfig"
+
 endmenu
Index: linux-2.6.38.8/drivers/Makefile
===================================================================
--- linux-2.6.38.8.orig/drivers/Makefile	2011-10-20 18:48:13.000000000 +0200
+++ linux-2.6.38.8/drivers/Makefile	2011-10-20 18:49:22.000000000 +0200
@@ -93,6 +93,7 @@
 obj-$(CONFIG_CPU_FREQ)		+= cpufreq/
 obj-$(CONFIG_CPU_IDLE)		+= cpuidle/
 obj-$(CONFIG_DMA_ENGINE)	+= dma/
+obj-$(CONFIG_ARCH_MXC)		+= mxc/
 obj-$(CONFIG_MMC)		+= mmc/
 obj-$(CONFIG_MEMSTICK)		+= memstick/
 obj-$(CONFIG_NEW_LEDS)		+= leds/
Index: linux-2.6.38.8/drivers/mxc/Kconfig
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.6.38.8/drivers/mxc/Kconfig	2011-10-20 18:51:18.000000000 +0200
@@ -0,0 +1,9 @@
+# drivers/video/Kconfig
+
+if ARCH_MXC
+
+menu "MXC support drivers"
+
+endmenu
+
+endif
