Allows to deactivate SDIO interrupt usage.

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>
Signed-off-by: Gwenhael Goavec-Merou <gwenhael.goavec-merou@armadeus.com>

Index: linux-3.8/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
===================================================================
--- linux-3.8.orig/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
+++ linux-3.8/Documentation/devicetree/bindings/mmc/fsl-imx-esdhc.txt
@@ -12,6 +12,7 @@ Required properties:
 Optional properties:
 - fsl,cd-controller : Indicate to use controller internal card detection
 - fsl,wp-controller : Indicate to use controller internal write protection
+- fsl,no_sdio_irq   : Indicate to not use SDIO IRQ
 
 Examples:
 
Index: linux-3.8/drivers/mmc/host/sdhci-esdhc-imx.c
===================================================================
--- linux-3.8.orig/drivers/mmc/host/sdhci-esdhc-imx.c
+++ linux-3.8/drivers/mmc/host/sdhci-esdhc-imx.c
@@ -421,6 +421,9 @@ sdhci_esdhc_imx_probe_dt(struct platform
 	if (of_get_property(np, "fsl,wp-controller", NULL))
 		boarddata->wp_type = ESDHC_WP_CONTROLLER;
 
+	if (of_get_property(np, "fsl,no_sdio_irq", NULL))
+		boarddata->no_sdio_irq = 1;
+
 	boarddata->cd_gpio = of_get_named_gpio(np, "cd-gpios", 0);
 	if (gpio_is_valid(boarddata->cd_gpio))
 		boarddata->cd_type = ESDHC_CD_GPIO;
@@ -575,6 +578,9 @@ static int sdhci_esdhc_imx_probe(struct
 		break;
 	}
 
+	if (boarddata->no_sdio_irq)
+		host->quirks2 = SDHCI_QUIRK2_HOST_OFF_CARD_ON;
+
 	err = sdhci_add_host(host);
 	if (err)
 		goto disable_clk;
Index: linux-3.8/drivers/mmc/host/sdhci.c
===================================================================
--- linux-3.8.orig/drivers/mmc/host/sdhci.c
+++ linux-3.8/drivers/mmc/host/sdhci.c
@@ -2823,7 +2823,10 @@ int sdhci_add_host(struct sdhci_host *ho
 
 	mmc->max_discard_to = (1 << 27) / host->timeout_clk;
 
-	mmc->caps |= MMC_CAP_SDIO_IRQ | MMC_CAP_ERASE | MMC_CAP_CMD23;
+	mmc->caps |= MMC_CAP_ERASE | MMC_CAP_CMD23;
+
+	if (!(host->quirks2 & SDHCI_QUIRK2_HOST_OFF_CARD_ON))
+		mmc->caps |= MMC_CAP_SDIO_IRQ;
 
 	if (host->quirks & SDHCI_QUIRK_MULTIBLOCK_READ_ACMD12)
 		host->flags |= SDHCI_AUTO_CMD12;
Index: linux-3.8/include/linux/platform_data/mmc-esdhc-imx.h
===================================================================
--- linux-3.8.orig/include/linux/platform_data/mmc-esdhc-imx.h
+++ linux-3.8/include/linux/platform_data/mmc-esdhc-imx.h
@@ -37,6 +37,7 @@ enum cd_types {
 struct esdhc_platform_data {
 	unsigned int wp_gpio;
 	unsigned int cd_gpio;
+	unsigned int no_sdio_irq;
 	enum wp_types wp_type;
 	enum cd_types cd_type;
 };
