Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: buildroot/package/owfs/Config.in
===================================================================
--- /dev/null
+++ buildroot/package/owfs/Config.in
@@ -0,0 +1,11 @@
+config BR2_PACKAGE_OWFS
+	bool "owfs"
+	select BR2_PACKAGE_LIBFUSE
+	help
+	  OWFS is an easy way to use the powerful 1-wire system of
+	  Dallas/Maxim.
+	  OWFS is a simple and flexible program to monitor and control the
+	  physical environment. You can write scripts to read temperature,
+	  flash lights, write to an LCD, log and graph, ...
+
+	  http://owfs.org/
Index: buildroot/package/owfs/owfs.mk
===================================================================
--- /dev/null
+++ buildroot/package/owfs/owfs.mk
@@ -0,0 +1,20 @@
+###############################################################################
+#
+# owfs
+#
+###############################################################################
+
+OWFS_SITE = http://downloads.sourceforge.net/owfs/
+OWFS_VERSION = 3.1p0
+OWFS_DEPENDENCIES = libfuse
+OWFS_CONF_OPTS += --disable-swig
+OWFS_INSTALL_STAGING = YES
+
+ifeq ($(BR2_PACKAGE_LIBUSB),y)
+OWFS_DEPENDENCIES += libusb
+OWFS_CONF_OPTS += --enable-usb
+else
+OWFS_CONF_OPTS += --disable-usb
+endif
+
+$(eval $(autotools-package))
Index: buildroot/package/Config.in
===================================================================
--- buildroot.orig/package/Config.in
+++ buildroot/package/Config.in
@@ -357,6 +357,7 @@ endif
 	source "package/open2300/Config.in"
 	source "package/openocd/Config.in"
 	source "package/openpowerlink/Config.in"
+	source "package/owfs/Config.in"
 	source "package/owl-linux/Config.in"
 	source "package/parted/Config.in"
 	source "package/pciutils/Config.in"
