Index: u-boot-2010.03/drivers/net/fec_mxc.c
===================================================================
--- u-boot-2010.03.orig/drivers/net/fec_mxc.c	2011-01-05 18:05:12.000000000 +0100
+++ u-boot-2010.03/drivers/net/fec_mxc.c	2011-01-05 18:10:35.000000000 +0100
@@ -325,7 +325,7 @@
 	for (i = 0; i < 6; i++)
 		mac[6-1-i] = readl(&iim->iim_bank_area0[IIM0_MAC + i]);
 
-	return is_valid_ether_addr(mac);
+	return !is_valid_ether_addr(mac);
 #endif
 }
 
