Index: u-boot-2010.03/MAKEALL
===================================================================
--- u-boot-2010.03.orig/MAKEALL	2011-01-05 18:05:14.000000000 +0100
+++ u-boot-2010.03/MAKEALL	2011-01-05 18:07:22.000000000 +0100
@@ -542,6 +542,7 @@
 
 LIST_ARM9="			\
 	a320evb			\
+	apf9328			\
 	ap920t			\
 	ap922_XA10		\
 	ap926ejs		\
Index: u-boot-2010.03/Makefile
===================================================================
--- u-boot-2010.03.orig/Makefile	2011-01-05 18:05:14.000000000 +0100
+++ u-boot-2010.03/Makefile	2011-01-05 18:07:22.000000000 +0100
@@ -2686,6 +2686,9 @@
 ## Atmel AT91RM9200 Systems
 #########################################################################
 
+apf9328_config		:	unconfig
+	@$(MKCONFIG) $(@:_config=) arm arm920t apf9328 armadeus imx
+
 at91rm9200dk_config	:	unconfig
 	@$(MKCONFIG) $(@:_config=) arm arm920t at91rm9200dk atmel at91rm9200
 
