Index: u-boot-2010.03/fs/jffs2/jffs2_1pass.c
===================================================================
--- u-boot-2010.03.orig/fs/jffs2/jffs2_1pass.c	2011-01-05 18:05:12.000000000 +0100
+++ u-boot-2010.03/fs/jffs2/jffs2_1pass.c	2011-01-05 18:10:39.000000000 +0100
@@ -196,8 +196,8 @@
 			if (nand_read(&nand_info[id->num], nand_cache_off,
 						&retlen, nand_cache) != 0 ||
 					retlen != NAND_CACHE_SIZE) {
-				printf("read_nand_cached: error reading nand off %#x size %d bytes\n",
-						nand_cache_off, NAND_CACHE_SIZE);
+/*				printf("read_nand_cached: error reading nand off %#x size %d bytes\n",
+						nand_cache_off, NAND_CACHE_SIZE);*/
 				return -1;
 			}
 		}
