Index: u-boot-2010.03/drivers/rtc/ds1374.c
===================================================================
--- u-boot-2010.03.orig/drivers/rtc/ds1374.c	2011-01-05 18:05:14.000000000 +0100
+++ u-boot-2010.03/drivers/rtc/ds1374.c	2011-01-05 18:08:28.000000000 +0100
@@ -113,6 +113,11 @@
 	unsigned int limit;
 	unsigned char tmp;
 	unsigned int i;
+	int old_bus;
+
+	/* switch to correct I2C bus */
+	old_bus = I2C_GET_BUS();
+	I2C_SET_BUS(CONFIG_SYS_RTC_BUS_NUM);
 
 	/*
 	 * Since the reads are being performed one byte at a time,
@@ -154,6 +159,9 @@
 		tm->tm_year, tm->tm_mon, tm->tm_mday, tm->tm_wday,
 		tm->tm_hour, tm->tm_min, tm->tm_sec);
 
+	/* switch back to original I2C bus */
+	I2C_SET_BUS(old_bus);
+
 	return rel;
 }
 
@@ -164,6 +172,11 @@
 
 	unsigned long time;
 	unsigned i;
+	int old_bus;
+
+	/* switch to correct I2C bus */
+	old_bus = I2C_GET_BUS();
+	I2C_SET_BUS(CONFIG_SYS_RTC_BUS_NUM);
 
 	DEBUGR ("Set DATE: %4d-%02d-%02d (wday=%d)  TIME: %2d:%02d:%02d\n",
 		tmp->tm_year, tmp->tm_mon, tmp->tm_mday, tmp->tm_wday,
@@ -187,6 +200,9 @@
 	/* Start clock */
 	rtc_write(RTC_CTL_ADDR, RTC_CTL_BIT_EN_OSC, FALSE);
 
+	/* switch back to original I2C bus */
+	I2C_SET_BUS(old_bus);
+
 	return 0;
 }
 
@@ -200,6 +216,11 @@
 void rtc_reset (void){
 
 	struct rtc_time tmp;
+	int old_bus;
+
+	/* switch to correct I2C bus */
+	old_bus = I2C_GET_BUS();
+	I2C_SET_BUS(CONFIG_SYS_RTC_BUS_NUM);
 
 	/* clear status flags */
 	rtc_write (RTC_SR_ADDR, (RTC_SR_BIT_AF|RTC_SR_BIT_OSF), FALSE); /* clearing OSF and AF */
@@ -232,6 +253,9 @@
 	rtc_write(RTC_WD_ALM_CNT_BYTE2_ADDR,0xAC, TRUE);
 	rtc_write(RTC_WD_ALM_CNT_BYTE1_ADDR,0xDE, TRUE);
 	rtc_write(RTC_WD_ALM_CNT_BYTE2_ADDR,0xAD, TRUE);
+
+	/* switch back to original I2C bus */
+	I2C_SET_BUS(old_bus);
 }
 
 /*
