Index: u-boot-2010.03/drivers/mtd/nand/nand_util.c
===================================================================
--- u-boot-2010.03.orig/drivers/mtd/nand/nand_util.c	2011-01-05 18:05:12.000000000 +0100
+++ u-boot-2010.03/drivers/mtd/nand/nand_util.c	2011-01-05 18:10:16.000000000 +0100
@@ -478,8 +478,7 @@
 	u_char *p_buffer = buffer;
 
 	/* Reject writes, which are not page aligned */
-	if ((offset & (nand->writesize - 1)) != 0 ||
-	    (*length & (nand->writesize - 1)) != 0) {
+	if ((offset & (nand->writesize - 1)) != 0 ) {
 		printf ("Attempt to write non page aligned data\n");
 		return -EINVAL;
 	}
Index: u-boot-2010.03/drivers/mtd/nand/nand_base.c
===================================================================
--- u-boot-2010.03.orig/drivers/mtd/nand/nand_base.c	2011-01-05 18:05:12.000000000 +0100
+++ u-boot-2010.03/drivers/mtd/nand/nand_base.c	2011-01-05 18:10:16.000000000 +0100
@@ -2002,7 +2002,7 @@
 		return 0;
 
 	/* reject writes, which are not page aligned */
-	if (NOTALIGNED(to) || NOTALIGNED(ops->len)) {
+	if (NOTALIGNED(to)) {
 		printk(KERN_NOTICE "nand_write: "
 		       "Attempt to write not page aligned data\n");
 		return -EINVAL;
