Add armadeus specific packages menu to BR.

Index: buildroot/package/Config.in
===================================================================
--- buildroot.orig/package/Config.in	2010-06-23 14:06:51.000000000 +0200
+++ buildroot/package/Config.in	2010-06-23 14:09:19.000000000 +0200
@@ -3,6 +3,9 @@
 source "package/busybox/Config.in"
 source "package/customize/Config.in"
 
+# Armadeus specific
+source "package/armadeus/Config.in"
+
 # Audio and video applications
 source "package/multimedia/Config.in"
 
