Signed-off-by: Eric Jarrige <eric.jarrige@armadeus.org>

Index: buildroot/boot/u-boot/Config.in
===================================================================
--- buildroot.orig/boot/u-boot/Config.in	2011-08-08 12:13:33.000000000 +0200
+++ buildroot/boot/u-boot/Config.in	2011-08-08 12:13:36.000000000 +0200
@@ -6,6 +6,7 @@
 if BR2_TARGET_UBOOT
 config BR2_TARGET_UBOOT_BOARDNAME
 	string "U-Boot board name"
+	default "$(BR2_BOARD_NAME)"
 	help
 	  One of U-Boot supported boards to be built.
 	  This will be suffixed with _config to meet U-Boot standard naming.
