Specific for Armadeus boards.

Index: buildroot/boot/u-boot/Config.in
===================================================================
--- buildroot.orig/boot/u-boot/Config.in	2011-08-08 12:13:43.000000000 +0200
+++ buildroot/boot/u-boot/Config.in	2011-08-08 12:13:56.000000000 +0200
@@ -25,6 +25,9 @@
 	help
 	  Select the specific U-Boot version you want to use
 
+config BR2_TARGET_UBOOT_1_3_4
+	bool "1.3.4"
+
 config BR2_TARGET_UBOOT_2011_03
 	bool "2011.03"
 
@@ -54,6 +57,7 @@
 
 config BR2_TARGET_UBOOT_VERSION
 	string
+	default "1.3.4"		if BR2_TARGET_UBOOT_1_3_4
 	default "2011.03"	if BR2_TARGET_UBOOT_2011_03
 	default "2010.12"	if BR2_TARGET_UBOOT_2010_12
 	default "2010.09"	if BR2_TARGET_UBOOT_2010_09
Index: buildroot/boot/u-boot/u-boot.mk
===================================================================
--- buildroot.orig/boot/u-boot/u-boot.mk	2011-08-08 12:13:43.000000000 +0200
+++ buildroot/boot/u-boot/u-boot.mk	2011-08-08 12:13:56.000000000 +0200
@@ -171,6 +171,11 @@
 # Build manually mkimage for the target
 $(TARGET_DIR)/usr/bin/mkimage: $(U_BOOT_DIR)/.configured
 	mkdir -p $(@D)
+ifeq ($(U_BOOT_VERSION),1.3.4)
+	$(TARGET_CC) -I$(U_BOOT_DIR)/include -DUSE_HOSTCC -o $@ \
+		$(U_BOOT_DIR)/tools/mkimage.c \
+		$(addprefix $(U_BOOT_DIR)/tools/,crc32.c image.c md5.c sha1.c fdt*.c)
+else
 	$(TARGET_CC) -I$(U_BOOT_DIR)/include -I$(U_BOOT_DIR)/tools \
 		-DUSE_HOSTCC -o $@ \
 		$(U_BOOT_DIR)/common/image.c \
@@ -179,6 +184,7 @@
 		$(addprefix $(U_BOOT_DIR)/lib*/,crc32.c md5.c sha1.c) \
 		$(U_BOOT_DIR)/tools/os_support.c \
 		$(wildcard $(U_BOOT_DIR)/libfdt/fdt*.c $(U_BOOT_DIR)/lib/libfdt/fdt*.c)
+endif
 
 # Build manually fw_printenv for the target
 $(TARGET_DIR)/usr/sbin/fw_printenv: $(U_BOOT_DIR)/.configured
