Index: buildroot/package/multimedia/Config.in
===================================================================
--- buildroot.orig/package/multimedia/Config.in	2011-01-27 15:24:49.000000000 +0100
+++ buildroot/package/multimedia/Config.in	2011-01-27 15:25:10.000000000 +0100
@@ -36,6 +36,7 @@
 source "package/multimedia/mpd/Config.in"
 source "package/multimedia/mpg123/Config.in"
 source "package/multimedia/mplayer/Config.in"
+source "package/pywebradio/Config.in"
 source "package/multimedia/musepack/Config.in"
 source "package/multimedia/python-mad/Config.in"
 source "package/multimedia/speex/Config.in"
Index: buildroot/package/pywebradio/Config.in
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/pywebradio/Config.in	2011-01-27 15:25:10.000000000 +0100
@@ -0,0 +1,6 @@
+config BR2_PACKAGE_PYWEBRADIO
+	bool "pywebradio"
+	help
+	  Play web radios
+
+	  http://www.pystream.info/
Index: buildroot/package/pywebradio/pywebradio.mk
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/pywebradio/pywebradio.mk	2011-01-27 15:25:10.000000000 +0100
@@ -0,0 +1,28 @@
+#############################################################
+#
+# pywebradio
+#
+#############################################################
+
+PYWEBRADIO_TRUNK:=http://forge.doumenc.org/svn/pyqtradio/trunk
+PYWEBRADIO_SITE:=
+PYWEBRADIO_VERSION:=421
+PYWEBRADIO_NAME:=pywebradio-svn-r$(PYWEBRADIO_VERSION)
+PYWEBRADIO_DIR:=$(BUILD_DIR)/$(PYWEBRADIO_NAME)
+PYWEBRADIO_SOURCE:=$(PYWEBRADIO_NAME).tar.bz2
+
+# GENTARGETS do not allow to get sources from SVN ??
+#$(DL_DIR)/$(PYWEBRADIO_SOURCE):
+#	$(SVN_CO) -r $(PYWEBRADIO_VERSION) $(PYWEBRADIO_TRUNK) $(PYWEBRADIO_DIR)
+#	tar -cv -C $(BUILD_DIR) $(PYWEBRADIO_NAME) | bzip2 - -c > $@
+
+define PYWEBRADIO_INSTALL_TARGET_CMDS
+	cd $(@D); \
+	python install.py -v -i pygame \
+		-s /usr/share/radio \
+		-c /etc/radio \
+		-e /usr/bin  \
+		-p $(TARGET_DIR)
+endef
+
+$(eval $(call GENTARGETS,package,pywebradio))
