Index: buildroot/package/Makefile.package.in
===================================================================
--- buildroot.orig/package/Makefile.package.in	2011-08-04 10:09:45.000000000 +0200
+++ buildroot/package/Makefile.package.in	2011-08-04 11:16:13.000000000 +0200
@@ -380,14 +380,16 @@
 # sanitize the package version that is used in paths, directory and file names.
 # Forward slashes may appear in the package's version when pointing to a
 # version control system branch or tag, for example remotes/origin/1_10_stable.
-$(2)_DL_VERSION	= $($(2)_VERSION)
 ifndef $(2)_VERSION
  ifdef $(3)_VERSION
-  $(2)_VERSION = $($(3)_VERSION)
+  $(2)_DL_VERSION = $($(3)_VERSION)
+  $(2)_VERSION = $(subst /,_,$($(3)_VERSION))
  else
+  $(2)_DL_VERSION = undefined
   $(2)_VERSION = undefined
  endif
 else
+ $(2)_DL_VERSION	= $($(2)_VERSION)
  $(2)_VERSION = $(subst /,_,$($(2)_VERSION))
 endif
 
