Temporary.

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: buildroot/package/busybox/Config.in
===================================================================
--- buildroot.orig/package/busybox/Config.in	2011-10-14 17:52:19.000000000 +0200
+++ buildroot/package/busybox/Config.in	2011-10-14 17:52:37.000000000 +0200
@@ -40,7 +40,7 @@
 	default "1.15.3"	if BR2_BUSYBOX_VERSION_1_15_X
 	default "1.16.2"	if BR2_BUSYBOX_VERSION_1_16_X
 	default "1.17.4"	if BR2_BUSYBOX_VERSION_1_17_X
-	default "1.18.4"	if BR2_BUSYBOX_VERSION_1_18_X
+	default "1.18.5"	if BR2_BUSYBOX_VERSION_1_18_X
 
 config BR2_PACKAGE_BUSYBOX_CONFIG
 	string "BusyBox configuration file to use?"
