Adding nanocom terminal

Signed-off-by: Fabien Marteau <fabien.marteau@armadeus.com>

Index: buildroot/package/nanocom/nanocom.mk
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/nanocom/nanocom.mk	2011-11-17 10:02:59.000000000 +0100
@@ -0,0 +1,26 @@
+#############################################################
+#
+# nanocom
+#
+#############################################################
+NANOCOM_VERSION = 1.0
+NANOCOM_SOURCE = nanocom.tar.gz
+NANOCOM_SITE:=http://$(BR2_SOURCEFORGE_MIRROR).dl.sourceforge.net/sourceforge/nanocom/
+
+define NANOCOM_BUILD_CMDS
+	$(MAKE) CC="$(TARGET_CC)" LD=$(TARGET_LD) -C $(@D) all
+endef
+
+define NANOCOM_INSTALL_TARGET_CMDS
+	$(INSTALL) -D -m 0755 $(@D)/nanocom $(TARGET_DIR)/usr/bin/nanocom
+endef
+
+# The .tar.bz2 is malformed (files aren't under a base directory)
+define NANOCOM_EXTRACT_TAR
+	@$(call MESSAGE, "Post extract")
+	$(TAR) -xf $(DL_DIR)/$($(PKG)_SOURCE) -C $(@D)
+endef
+
+NANOCOM_POST_EXTRACT_HOOKS += NANOCOM_EXTRACT_TAR
+
+$(eval $(call GENTARGETS, package, nanocom))
Index: buildroot/package/Config.in
===================================================================
--- buildroot.orig/package/Config.in	2011-11-17 10:00:48.000000000 +0100
+++ buildroot/package/Config.in	2011-11-17 10:00:48.000000000 +0100
@@ -211,6 +211,7 @@
 source "package/mdadm/Config.in"
 source "package/memtester/Config.in"
 source "package/mtd/Config.in"
+source "package/nanocom/Config.in"
 source "package/ntfs-3g/Config.in"
 source "package/ntfsprogs/Config.in"
 source "package/pciutils/Config.in"
Index: buildroot/package/nanocom/Config.in
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/nanocom/Config.in	2011-11-17 10:03:35.000000000 +0100
@@ -0,0 +1,5 @@
+config BR2_PACKAGE_NANOCOM
+	bool "nanocom"
+	help
+	  Simple terminal communication tool
+
