Temporary fix until mplayer fixes the cortex-A8 optimized build.

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: buildroot/package/multimedia/mplayer/mplayer.mk
===================================================================
--- buildroot.orig/package/multimedia/mplayer/mplayer.mk	2012-01-13 14:57:49.000000000 +0100
+++ buildroot/package/multimedia/mplayer/mplayer.mk	2012-01-16 19:00:03.000000000 +0100
@@ -91,7 +91,7 @@
 endif
 
 ifeq ($(call qstrip,$(BR2_GCC_TARGET_ARCH)),armv7-a)
-MPLAYER_CONF_OPTS += --enable-neon
+MPLAYER_CONF_OPTS += --enable-armv6
 endif
 
 define MPLAYER_CONFIGURE_CMDS
