Add bluez package (3.36) to BR.

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: buildroot/package/Config.in
===================================================================
--- buildroot.orig/package/Config.in	2012-02-21 16:44:52.000000000 +0100
+++ buildroot/package/Config.in	2012-02-21 16:44:54.000000000 +0100
@@ -683,6 +683,8 @@
 source "package/bcusdk/Config.in"
 source "package/bind/Config.in"
 source "package/bluez_utils/Config.in"
+source "package/bluez3-libs/Config.in"
+source "package/bluez3-utils/Config.in"
 source "package/bmon/Config.in"
 source "package/boa/Config.in"
 source "package/bridge-utils/Config.in"
Index: buildroot/package/bluez3-libs/Config.in
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/bluez3-libs/Config.in	2012-02-21 16:44:54.000000000 +0100
@@ -0,0 +1,6 @@
+config BR2_PACKAGE_BLUEZ3_LIBS
+	bool "bluez (v3)"
+	select BR2_PACKAGE_PKGCONFIG
+	help
+	  Install the bluez v3 library
+
Index: buildroot/package/bluez3-libs/bluez3-libs.mk
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/bluez3-libs/bluez3-libs.mk	2012-02-21 16:44:54.000000000 +0100
@@ -0,0 +1,13 @@
+#############################################################
+#
+# bluez-libs
+#
+#############################################################
+BLUEZ3_LIBS_VERSION:=3.36
+BLUEZ3_LIBS_SITE = http://bluez.sf.net/download
+BLUEZ3_LIBS_SOURCE = bluez-libs-$(BLUEZ3_LIBS_VERSION).tar.gz
+BLUEZ3_LIBS_INSTALL_STAGING = YES
+BLUEZ3_LIBS_INSTALL_TARGET = YES
+
+$(eval $(autotools-package))
+
Index: buildroot/package/bluez3-utils/Config.in
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/bluez3-utils/Config.in	2012-02-21 16:44:54.000000000 +0100
@@ -0,0 +1,9 @@
+config BR2_PACKAGE_BLUEZ3_UTILS
+	bool "bluez-utils (v3)"
+	depends on BR2_PACKAGE_DBUS
+	select BR2_PACKAGE_BLUEZ3_LIBS
+	help
+	  Install the bluez v3 utilities
+
+comment "bluez-utils not available (need dbus)"
+        depends on !BR2_PACKAGE_DBUS
Index: buildroot/package/bluez3-utils/bluez3-utils.mk
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/bluez3-utils/bluez3-utils.mk	2012-02-21 16:44:54.000000000 +0100
@@ -0,0 +1,53 @@
+#############################################################
+#
+# bluez-utils - User space programs for controling Bluetooth
+#
+#############################################################
+BLUEZ3_UTILS_VERSION:=3.36
+BLUEZ3_UTILS_SITE = http://bluez.sf.net/download
+BLUEZ3_UTILS_SOURCE = bluez-utils-$(BLUEZ3_UTILS_VERSION).tar.gz
+
+ifeq ($(strip $(BR2_PACKAGE_ALSA-LIB)),y)
+ALSA_DEP=alsa
+WITH_ALSA=--with-alsa=$(STAGING_DIR)
+else
+WITH_ALSA=--without-alsa
+endif
+
+ifeq ($(strip $(BR2_PACKAGE_OPENOBEX)),y)
+OBEX_DEP=openobex
+WITH_OBEX=--with-openobex=$(STAGING_DIR)
+else
+OBEX_DEP=
+WITH_OBEX=--without-obex
+endif
+
+BLUEZ3_UTILS_CONF_OPT = \
+		$(WITH_OBEX) \
+		$(WITH_ALSA) \
+		--disable-pie \
+		--disable-expat \
+		--enable-network \
+		--enable-serial \
+		--enable-audio \
+		--enable-input \
+		--enable-sync \
+		--enable-hcid \
+		--enable-hidd \
+		--enable-pand \
+		--disable-test \
+		--enable-configfiles \
+		--disable-manpages \
+		--enable-initscripts \
+		--disable-pcmciarules \
+		--disable-bccmd \
+		--enable-avctrl \
+		--enable-hid2hci \
+		--disable-dfutool
+
+# --disable-pie: uClibc doesn't handle "position independent executables" (segfault at launching)
+
+BLUEZ3_UTILS_DEPENDENCIES = bluez3-libs dbus $(OBEX_DEP) $(ALSA_DEP)
+
+$(eval $(autotools-package))
+
