BusyBox: Fix daily snapshot config file

Signed-off-by: Janos Angeli <angelo@angelo.hu>
---

Index: buildroot/package/busybox/Config.in
===================================================================
--- buildroot.orig/package/busybox/Config.in
+++ buildroot/package/busybox/Config.in
@@ -33,6 +33,7 @@ endchoice
 
 config BR2_BUSYBOX_VERSION
 	string
+	default "snapshot"	if BR2_PACKAGE_BUSYBOX_SNAPSHOT
 	default "1.19.4"	if BR2_BUSYBOX_VERSION_1_19_X
 	default "1.20.2"	if BR2_BUSYBOX_VERSION_1_20_X
 	default "1.21.0"	if BR2_BUSYBOX_VERSION_1_21_X
