
Signed-off-by: Gwenhael Goavec-Merou <gwenhael.goavec-merou@armadeus.com>
---
Index: linux-3.12-rc1/arch/arm/boot/dts/imx27.dtsi
===================================================================
--- linux-3.12-rc1.orig/arch/arm/boot/dts/imx27.dtsi
+++ linux-3.12-rc1/arch/arm/boot/dts/imx27.dtsi
@@ -526,6 +526,40 @@
 		};
 	};
 
+	imxfb {
+		pinctrl_imxfb1_1: imxfb1-1 {
+			fsl,pins = <
+				MX27_PAD_LSCLK__LSCLK 0x0
+				MX27_PAD_LD0__LD0 0x0
+				MX27_PAD_LD1__LD1 0x0
+				MX27_PAD_LD2__LD2 0x0
+				MX27_PAD_LD3__LD3 0x0
+				MX27_PAD_LD4__LD4 0x0
+				MX27_PAD_LD5__LD5 0x0
+				MX27_PAD_LD6__LD6 0x0
+				MX27_PAD_LD7__LD7 0x0
+				MX27_PAD_LD8__LD8 0x0
+				MX27_PAD_LD9__LD9 0x0
+				MX27_PAD_LD10__LD10 0x0
+				MX27_PAD_LD11__LD11 0x0
+				MX27_PAD_LD12__LD12 0x0
+				MX27_PAD_LD13__LD13 0x0
+				MX27_PAD_LD14__LD14 0x0
+				MX27_PAD_LD15__LD15 0x0
+				MX27_PAD_LD16__LD16 0x0
+				MX27_PAD_LD17__LD17 0x0
+				MX27_PAD_REV__REV 0x0
+				MX27_PAD_CLS__CLS 0x0
+				MX27_PAD_PS__PS 0x0
+				MX27_PAD_SPL_SPR__SPL_SPR 0x0
+				MX27_PAD_HSYNC__HSYNC 0x0
+				MX27_PAD_VSYNC__VSYNC 0x0
+				MX27_PAD_CONTRAST__CONTRAST 0x0
+				MX27_PAD_OE_ACD__OE_ACD 0x0
+			>;
+		};
+	};
+
 	owire {
 		pinctrl_owire1_1: owire1-1 {
 			fsl,pins = <
