Add apf51 basic support

Signed-off-by: Laurent Cans <laurent.cans@gmail.com>
Signed-off-by: Gwenhael Goavec-Merou <gwenhael.goavec-merou@armadeus.com>
---
Index: linux-3.13-rc2/arch/arm/boot/dts/imx51-apf51.dts
===================================================================
--- linux-3.13-rc2.orig/arch/arm/boot/dts/imx51-apf51.dts
+++ linux-3.13-rc2/arch/arm/boot/dts/imx51-apf51.dts
@@ -30,6 +30,19 @@
 			clock-frequency = <33554432>;
 		};
 	};
+
+	regulators {
+		compatible = "simple-bus";
+
+		reg_usbotg_vbus: usbotg_vbus {
+			compatible = "regulator-fixed";
+			regulator-name = "usbotg_vbus";
+			regulator-min-microvolt = <5000000>;
+			regulator-max-microvolt = <5000000>;
+			gpio = <&gpio1 8 0>;
+			enable-active-high;
+		};
+	};
 };
 
 &fec {
@@ -53,3 +66,31 @@
 	pinctrl-0 = <&pinctrl_uart3_2>;
 	status = "okay";
 };
+
+&usbphy1 {
+	reset-gpios = <&gpio3 1 1>;
+};
+
+&usbh1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usbh1_1>;
+	phy_type = "ulpi";
+	dr_mode = "host";
+	ext-clk-en;
+	status = "okay";
+};
+
+&usbh2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_usbh2_1>;
+	phy_type = "ulpi";
+	dr_mode = "host";
+	ext-clk-en;
+	status = "okay";
+};
+
+&usbotg {
+	phy_type = "utmi_wide";
+	vbus-supply = <&reg_usbotg_vbus>;
+	status = "okay";
+};
