
Signed-off-by: Gwenhael Goavec-Merou <gwenhael.goavec-merou@armadeus.com>
---
Index: linux-3.12-rc1/drivers/usb/chipidea/ci_hdrc_imx.c
===================================================================
--- linux-3.12-rc1.orig/drivers/usb/chipidea/ci_hdrc_imx.c
+++ linux-3.12-rc1/drivers/usb/chipidea/ci_hdrc_imx.c
@@ -67,6 +67,9 @@ static struct imx_usbmisc_data *usbmisc_
 	if (of_find_property(np, "external-vbus-divider", NULL))
 		data->evdo = 1;
 
+	if (!of_find_property(np, "ext-clk-en", NULL))
+		data->ext_clk_en = 1;
+
 	return data;
 }
 
Index: linux-3.12-rc1/drivers/usb/chipidea/ci_hdrc_imx.h
===================================================================
--- linux-3.12-rc1.orig/drivers/usb/chipidea/ci_hdrc_imx.h
+++ linux-3.12-rc1/drivers/usb/chipidea/ci_hdrc_imx.h
@@ -14,6 +14,7 @@ struct imx_usbmisc_data {
 
 	unsigned int disable_oc:1; /* over current detect disabled */
 	unsigned int evdo:1; /* set external vbus divider option */
+	unsigned int ext_clk_en:1; /* select the clock from external PHY*/
 };
 
 int imx_usbmisc_init(struct imx_usbmisc_data *);
