Add support for eMMC 5.0.

Signed-off-by: Sébastien Szymanski <sebastien.szymanski@armadeus.com>
Index: linux-wandboard_imx_3.10.31_1.1.0_beta2/drivers/mmc/core/mmc.c
===================================================================
--- linux-wandboard_imx_3.10.31_1.1.0_beta2.orig/drivers/mmc/core/mmc.c
+++ linux-wandboard_imx_3.10.31_1.1.0_beta2/drivers/mmc/core/mmc.c
@@ -293,12 +293,6 @@ static int mmc_read_ext_csd(struct mmc_c
	}

	card->ext_csd.rev = ext_csd[EXT_CSD_REV];
-	if (card->ext_csd.rev > 6) {
-		pr_err("%s: unrecognised EXT_CSD revision %d\n",
-			mmc_hostname(card->host), card->ext_csd.rev);
-		err = -EINVAL;
-		goto out;
-	}

	card->ext_csd.raw_sectors[0] = ext_csd[EXT_CSD_SEC_CNT + 0];
	card->ext_csd.raw_sectors[1] = ext_csd[EXT_CSD_SEC_CNT + 1];
