Bump setserial patch version.

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: buildroot/package/setserial/setserial.mk
===================================================================
--- buildroot.orig/package/setserial/setserial.mk	2012-02-21 16:43:24.000000000 +0100
+++ buildroot/package/setserial/setserial.mk	2012-02-21 16:46:07.000000000 +0100
@@ -4,7 +4,7 @@
 #
 #############################################################
 SETSERIAL_VERSION            = 2.17
-SETSERIAL_PATCH              = setserial_2.17-45.2.diff.gz
+SETSERIAL_PATCH              = setserial_2.17-45.3.diff.gz
 SETSERIAL_SOURCE             = setserial_$(SETSERIAL_VERSION).orig.tar.gz
 SETSERIAL_SITE               = $(BR2_DEBIAN_MIRROR)/debian/pool/main/s/setserial/
 
