Bump to tslib 1.1 to fix compilation error on Ubuntu 13.10

Signed-off-by: Eric Jarrige <eric.jarrige@armadeus.org>
---

--- buildroot/package/tslib/tslib.mk
+++ buildroot/package/tslib/tslib.mk
@@ -3,8 +3,8 @@
 # tslib
 #
 #############################################################
-TSLIB_VERSION = 158ee49b32f83cb7b02d5315f41c2e4cff38942d
-TSLIB_SITE = http://github.com/kergoth/tslib/tarball/master
+TSLIB_VERSION = 1.1
+TSLIB_SITE = http://github.com/kergoth/tslib/tarball/$(TSLIB_VERSION)
 TSLIB_LICENSE = GPL, LGPL
 TSLIB_LICENSE_FILES = COPYING
 
