Index: uboot-2014.07/board/armadeus/apf6/apf6.c
===================================================================
--- uboot-2014.07.orig/board/armadeus/apf6/apf6.c
+++ uboot-2014.07/board/armadeus/apf6/apf6.c
@@ -18,6 +18,7 @@
 #include <asm/imx-common/video.h>
 #include <asm/io.h>
 #include <common.h>
+#include <environment.h>
 #include <fsl_esdhc.h>
 #include <linux/ctype.h>
 #include <miiphy.h>
@@ -327,9 +328,17 @@ int board_late_init(void)
 	u_char *firmware_buffer =
 	    (u_char *) (CONFIG_SYS_LOAD_ADDR + CONFIG_SYS_MONITOR_LEN);
 	size_t size = 0;
+	struct src *psrc = (struct src *)SRC_BASE_ADDR;
+	unsigned reg = readl(&psrc->sbmr2);
 
 	SETUP_IOMUX_PADS(usb_pads);
 
+	/* In bootstrap don't use the the env vars */
+	if (((reg & 0x3000000) >> 24) == 0x1) {
+		set_default_env(NULL);
+		setenv("preboot", "");
+	}
+
 	if (is_cpu_type(MXC_CPU_MX6Q) || is_cpu_type(MXC_CPU_MX6D))
 		setenv("fdt_file", "imx6q-apf6dev.dtb");
 	else if (is_cpu_type(MXC_CPU_MX6DL) || is_cpu_type(MXC_CPU_MX6SOLO))
