Index: linux-2.6.29.4/drivers/i2c/busses/Kconfig
===================================================================
--- linux-2.6.29.4.orig/drivers/i2c/busses/Kconfig	2009-06-04 10:07:14.000000000 +0200
+++ linux-2.6.29.4/drivers/i2c/busses/Kconfig	2009-06-04 10:37:46.000000000 +0200
@@ -628,6 +628,12 @@
 	  This driver can also be built as a module.  If so, the module
 	  will be called i2c-pca-isa.
 
+config I2C_IMX
+	tristate "i2c bus on i.MX"
+	depends on I2C
+	help
+	  i2c implementation on i.MX Freescale processors
+
 	  This device is almost undetectable and using this driver on a
 	  system which doesn't have this device will result in long
 	  delays when I2C/SMBus chip drivers are loaded (e.g. at boot
Index: linux-2.6.29.4/drivers/i2c/busses/Makefile
===================================================================
--- linux-2.6.29.4.orig/drivers/i2c/busses/Makefile	2009-06-04 10:07:14.000000000 +0200
+++ linux-2.6.29.4/drivers/i2c/busses/Makefile	2009-06-04 10:37:09.000000000 +0200
@@ -44,6 +44,7 @@
 obj-$(CONFIG_I2C_PXA)		+= i2c-pxa.o
 obj-$(CONFIG_I2C_S3C2410)	+= i2c-s3c2410.o
 obj-$(CONFIG_I2C_SH7760)	+= i2c-sh7760.o
+obj-$(CONFIG_I2C_IMX)		+= i2c-imx.o
 obj-$(CONFIG_I2C_SH_MOBILE)	+= i2c-sh_mobile.o
 obj-$(CONFIG_I2C_SIMTEC)	+= i2c-simtec.o
 obj-$(CONFIG_I2C_VERSATILE)	+= i2c-versatile.o
