--- linux-2.6.27.2.org/arch/arm/mach-imx/clock.c	2008-10-23 09:17:23.000000000 +0200
+++ linux-2.6.27.2/arch/arm/mach-imx/clock.c	2008-10-23 09:19:43.000000000 +0200
@@ -107,6 +107,15 @@
 	return imx_get_system_clk() / (((CSCR>>10) & 0xf)+1);
 }
 
+/*
+ *  get USB clock
+ */
+static unsigned long imx_get_usb_clk(void)
+{
+	return imx_get_system_clk() / (((CSCR>>26) & 0x7)+1);
+}
+
+
 static struct clk clk_system_clk = {
 	.name = "system_clk",
 	.get_rate = imx_get_system_clk,
@@ -142,6 +151,11 @@
 	.get_rate = imx_get_perclk3,
 };
 
+static struct clk clk_usb_clk = {
+	.name = "usbd_clk",
+	.get_rate = imx_get_usb_clk,
+};
+
 static struct clk *clks[] = {
 	&clk_perclk1,
 	&clk_perclk2,
@@ -150,6 +164,7 @@
 	&clk_hclk,
 	&clk_mcu_clk,
 	&clk_uart_clk,
+	&clk_usb_clk,
 };
 
 static LIST_HEAD(clocks);
