Index: linux-2.6.29/drivers/usb/gadget/Kconfig
===================================================================
--- linux-2.6.29.orig/drivers/usb/gadget/Kconfig	2009-05-01 14:28:39.000000000 +0200
+++ linux-2.6.29/drivers/usb/gadget/Kconfig	2009-05-01 14:29:27.000000000 +0200
@@ -307,8 +307,8 @@
 	  the TI DaVinci, OMAP 243x, OMAP 343x, TUSB 6010, and ADI Blackfin
 
 config USB_GADGET_IMX
-	boolean "Freescale IMX USB Peripheral Controller"
-	depends on ARCH_MX1
+	boolean "Freescale i.MX1/L/S USB Peripheral Controller"
+	depends on ARCH_MX1 || ARCH_IMX
 	help
 	   Freescale's IMX series include an integrated full speed
 	   USB 1.1 device controller.  The controller in the IMX series
