Index: linux-2.6.29.4/arch/arm/mach-imx/include/mach/imxfb.h
===================================================================
--- linux-2.6.29.4.orig/arch/arm/mach-imx/include/mach/imxfb.h	2009-06-04 11:30:48.000000000 +0200
+++ linux-2.6.29.4/arch/arm/mach-imx/include/mach/imxfb.h	2009-06-04 11:30:49.000000000 +0200
@@ -84,3 +84,13 @@
 };
 
 void set_imx_fb_info(struct imx_fb_platform_data *);
+
+/* --- Backlight handling --- */
+struct imxbl_machinfo {
+	int max_intensity;
+	int default_intensity;
+	int limit_mask; /* When battery is low */
+	void (*set_bl_intensity)(int intensity);
+};
+extern void imxbl_limit_intensity(int limit);
+
