--- linux.org/scripts/checksyscalls.sh	2008-10-07 00:50:25.000000000 +0200
+++ linux/scripts/checksyscalls.sh	2008-10-07 00:49:39.000000000 +0200
@@ -29,6 +29,13 @@ cat << EOF
 #define __IGNORE_statfs64
 #endif
 
+#define __IGNORE_epoll_pwait
+#define __IGNORE_migrate_pages
+#define __IGNORE_pselect6
+#define __IGNORE_ppoll
+#define __IGNORE_fallocate
+#define __IGNORE_fadvise64
+
 /* i386-specific or historical system calls */
 #define __IGNORE_break
 #define __IGNORE_stty
