Needed for POD IRQ handler registering (module).

Index: linux-2.6.29/arch/arm/kernel/irq.c
===================================================================
--- linux-2.6.29.orig/arch/arm/kernel/irq.c	2009-03-24 00:12:14.000000000 +0100
+++ linux-2.6.29/arch/arm/kernel/irq.c	2009-04-02 13:25:16.000000000 +0200
@@ -152,6 +152,7 @@
 		desc->status &= ~IRQ_NOAUTOEN;
 	spin_unlock_irqrestore(&desc->lock, flags);
 }
+EXPORT_SYMBOL(set_irq_flags);
 
 void __init init_IRQ(void)
 {
Index: linux-2.6.29/arch/arm/mach-imx/include/mach/irqs.h
===================================================================
--- linux-2.6.29.orig/arch/arm/mach-imx/include/mach/irqs.h	2009-04-02 13:25:15.000000000 +0200
+++ linux-2.6.29/arch/arm/mach-imx/include/mach/irqs.h	2009-04-02 13:25:16.000000000 +0200
@@ -117,6 +117,10 @@
 /* switch betwean IRQ and FIQ */
 extern int imx_set_irq_fiq(unsigned int irq, unsigned int type);
 
-#define NR_IRQS (IRQ_GPIOD(32) + 1)
+#define IRQ_FPGA_START (IRQ_GPIOD(32) + 1)
+#define IRQ_FPGA(x) (IRQ_FPGA_START + x)
+
+//#define NR_IRQS (IRQ_GPIOD(32) + 1)
+#define NR_IRQS (IRQ_FPGA(16) + 1)
 #define IRQ_GPIO(x)
 #endif
Index: linux-2.6.29/kernel/irq/chip.c
===================================================================
--- linux-2.6.29.orig/kernel/irq/chip.c	2009-03-24 00:12:14.000000000 +0100
+++ linux-2.6.29/kernel/irq/chip.c	2009-04-02 13:28:48.000000000 +0200
@@ -516,6 +516,7 @@
 out_unlock:
 	spin_unlock(&desc->lock);
 }
+EXPORT_SYMBOL(handle_edge_irq);
 
 /**
  *	handle_percpu_IRQ - Per CPU local irq handler
@@ -603,6 +604,7 @@
 	set_irq_chip(irq, chip);
 	__set_irq_handler(irq, handle, 0, NULL);
 }
+EXPORT_SYMBOL(set_irq_chip_and_handler);
 
 void
 set_irq_chip_and_handler_name(unsigned int irq, struct irq_chip *chip,
Index: linux-2.6.29/kernel/irq/handle.c
===================================================================
--- linux-2.6.29.orig/kernel/irq/handle.c	2009-03-24 00:12:14.000000000 +0100
+++ linux-2.6.29/kernel/irq/handle.c	2009-04-02 13:30:22.000000000 +0200
@@ -226,6 +226,7 @@
 #endif
 	}
 };
+EXPORT_SYMBOL(irq_desc);
 
 int __init early_irq_init(void)
 {
