Index: linux-2.6.29/drivers/serial/imx.c
===================================================================
--- linux-2.6.29.orig/drivers/serial/imx.c	2009-03-24 00:12:14.000000000 +0100
+++ linux-2.6.29/drivers/serial/imx.c	2009-03-31 10:57:09.000000000 +0200
@@ -657,6 +657,18 @@
 	writel(temp, sport->port.membase + UCR1);
 }
 
+static void imx_serial_dump_registers(struct imx_port *sport)
+{
+	printk("UCR1: 0x%08x\n", readl(sport->port.membase + UCR1));
+	printk("UCR2: 0x%08x\n", readl(sport->port.membase + UCR2));
+	printk("UCR3: 0x%08x\n", readl(sport->port.membase + UCR3));
+	printk("UCR4: 0x%08x\n", readl(sport->port.membase + UCR4));
+	printk("UFCR: 0x%08x\n", readl(sport->port.membase + UFCR));
+	printk("UBIR: 0x%08x\n", readl(sport->port.membase + UBIR));
+	printk("UBMR: 0x%08x\n", readl(sport->port.membase + UBMR));
+	printk("UBRC: 0x%08x\n", readl(sport->port.membase + UBRC));
+}
+
 static void
 imx_set_termios(struct uart_port *port, struct ktermios *termios,
 		   struct ktermios *old)
