Index: linux-2.6.29.6/drivers/video/mxc/mxcfb_modedb.c
===================================================================
--- linux-2.6.29.6.orig/drivers/video/mxc/mxcfb_modedb.c	2009-10-11 19:44:36.000000000 +0200
+++ linux-2.6.29.6/drivers/video/mxc/mxcfb_modedb.c	2009-10-11 19:45:36.000000000 +0200
@@ -101,6 +101,19 @@
 		.vmode = FB_VMODE_NONINTERLACED,
 		.flag = 0,
 	},
+        {
+		.name = "Chimei-LW700AT9003",
+		.refresh = 60,
+		.xres = 800,
+		.yres = 480,
+		.pixclock = 30303, /* picoS */
+		.left_margin = 96, .right_margin = 96, /* at least 3 & 1 */
+		.upper_margin = 0x14, .lower_margin = 0x15,
+		.hsync_len = 64, .vsync_len = 4,
+		.sync = FB_SYNC_CLK_IDLE_EN | FB_SYNC_OE_ACT_HIGH | FB_SYNC_CLK_INVERT,
+		.vmode = FB_VMODE_NONINTERLACED,
+		.flag = 0,
+	},
 };
 
 int mxcfb_modedb_sz = ARRAY_SIZE(mxcfb_modedb);
