--- linux-2.6.27.2.ref/arch/arm/plat-mxc/include/mach/system.h	2008-12-18 17:46:40.000000000 +0100
+++ linux-2.6.27.2/arch/arm/plat-mxc/include/mach/system.h	2008-12-18 17:45:01.000000000 +0100
@@ -21,14 +21,9 @@
 #ifndef __ASM_ARCH_MXC_SYSTEM_H__
 #define __ASM_ARCH_MXC_SYSTEM_H__
 
-static inline void arch_idle(void)
-{
-	cpu_do_idle();
-}
+extern void arch_idle(void);
+
+extern void arch_reset(char mode);
 
-static inline void arch_reset(char mode)
-{
-	cpu_reset(0);
-}
 
 #endif /* __ASM_ARCH_MXC_SYSTEM_H__ */
