Increases max DMA zone size for MXC platform.

Signed-off-by: Nicolas Colombain <nicolas.colombain@armadeus.com>
Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: linux-2.6.29.4/arch/arm/plat-mxc/include/mach/memory.h
===================================================================
--- linux-2.6.29.4.orig/arch/arm/plat-mxc/include/mach/memory.h	2009-03-24 00:12:14.000000000 +0100
+++ linux-2.6.29.4/arch/arm/plat-mxc/include/mach/memory.h	2009-06-04 11:24:55.000000000 +0200
@@ -19,4 +19,7 @@
 #define PHYS_OFFSET		UL(0x80000000)
 #endif
 
+/* Override the ARM default */
+#define CONSISTENT_DMA_SIZE SZ_4M
+
 #endif /* __ASM_ARCH_MXC_MEMORY_H__ */
