Adds PWM ressources on MX2.

Signed-off-by: Nicolas Colombain <nicolas.colombain@armadeus.com>
Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: linux-2.6.29.4/arch/arm/mach-mx2/devices.c
===================================================================
--- linux-2.6.29.4.orig/arch/arm/mach-mx2/devices.c	2009-06-04 11:24:54.000000000 +0200
+++ linux-2.6.29.4/arch/arm/mach-mx2/devices.c	2009-06-04 11:29:59.000000000 +0200
@@ -657,6 +657,25 @@
 	.resource = mxc_ehci2_resources,
 };
 
+static struct resource mxc_pwm_resources[] = {
+    [0] = {
+        .start  = (PWM_BASE_ADDR),
+        .end    = (PWM_BASE_ADDR + 0x14),
+        .flags  = IORESOURCE_MEM,
+    },
+    [1] = {
+        .start  = (MXC_INT_PWM),
+        .end    = (MXC_INT_PWM),
+        .flags  = IORESOURCE_IRQ,
+    },
+};
+
+struct platform_device mxc_pwm_device0 = {
+    .name           = "imx-pwm",
+    .id             = 0,
+    .num_resources  = ARRAY_SIZE(mxc_pwm_resources),
+    .resource       = mxc_pwm_resources,
+};
 
 /* GPIO port description */
 static struct mxc_gpio_port imx_gpio_ports[] = {
Index: linux-2.6.29.4/arch/arm/mach-mx2/devices.h
===================================================================
--- linux-2.6.29.4.orig/arch/arm/mach-mx2/devices.h	2009-06-04 11:24:54.000000000 +0200
+++ linux-2.6.29.4/arch/arm/mach-mx2/devices.h	2009-06-04 11:24:55.000000000 +0200
@@ -30,3 +30,4 @@
 extern struct platform_device mxc_otg;
 extern struct platform_device mxc_ehci1;
 extern struct platform_device mxc_ehci2;
+extern struct platform_device mxc_pwm_device0;
