Updates drivers/spi/spi_imx.c to 2.6.32 new features.

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: linux-2.6.29.6/drivers/spi/spi_imx.c
===================================================================
--- linux-2.6.29.6.orig/drivers/spi/spi_imx.c	2010-01-19 14:30:17.000000000 +0100
+++ linux-2.6.29.6/drivers/spi/spi_imx.c	2010-01-19 14:33:50.000000000 +0100
@@ -1275,9 +1275,6 @@
 	return -EINVAL;
 }
 
-/* the spi->mode bits understood by this driver: */
-#define MODEBITS (SPI_CPOL | SPI_CPHA | SPI_CS_HIGH)
-
 /* On first setup bad values must free chip_data memory since will cause
    spi_new_device to fail. Bad value setup from protocol driver are simply not
    applied and notified to the calling driver. */
@@ -1290,12 +1287,6 @@
 	u32 tmp;
 	int status = 0;
 
-	if (spi->mode & ~MODEBITS) {
-		dev_dbg(&spi->dev, "setup: unsupported mode bits %x\n",
-			spi->mode & ~MODEBITS);
-		return -EINVAL;
-	}
-
 	/* Get controller data */
 	chip_info = spi->controller_data;
 
@@ -1590,6 +1581,7 @@
 	master->cleanup = cleanup;
 	master->setup = setup;
 	master->transfer = transfer;
+	master->mode_bits = SPI_CPOL | SPI_CPHA | SPI_CS_HIGH;
 
 	drv_data->dummy_dma_buf = SPI_DUMMY_u32;
 
