From 8262b3cd32259533556e6dd8d1cd0a4c77c6acfe Mon Sep 17 00:00:00 2001
From: Philippe Reynes <tremyfr@yahoo.fr>
Date: Mon, 15 Jul 2013 15:26:14 +0200
Subject: [PATCH] Fix spidev

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>
---
Index: linux-3.14-rc4/arch/arm/boot/dts/imx28-apf28dev.dts
===================================================================
--- linux-3.14-rc4.orig/arch/arm/boot/dts/imx28-apf28dev.dts
+++ linux-3.14-rc4/arch/arm/boot/dts/imx28-apf28dev.dts
@@ -32,6 +32,13 @@
 				pinctrl-names = "default";
 				pinctrl-0 = <&spi2_pins_a>;
 				status = "okay";
+/* Uncomment to have spidev on SPI2 (ChipSelect 0):
+				device@0 {
+					compatible = "spidev";
+					reg = <0>;
+					spi-max-frequency = <3125000>;
+				};
+*/
 			};
 
 			pinctrl@80018000 {
@@ -48,6 +55,8 @@
 						MX28_PAD_LCD_D20__GPIO_1_20
 						MX28_PAD_LCD_D21__GPIO_1_21
 						MX28_PAD_LCD_D22__GPIO_1_22
+						MX28_PAD_LCD_D23__GPIO_1_23
+						MX28_PAD_GPMI_CE1N__GPIO_0_17
 					>;
 					fsl,drive-strength = <MXS_DRIVE_4mA>;
 					fsl,voltage = <MXS_VOLTAGE_HIGH>;
@@ -66,6 +75,15 @@
 					fsl,voltage = <MXS_VOLTAGE_HIGH>;
 					fsl,pull-up = <MXS_PULL_DISABLE>;
 				};
+
+				apf28dev_usb0_id_pins: apf28dev-usb0-id-pins@0 {
+					reg = <0>;
+					fsl,pinmux-ids = <
+						MX28_PAD_PWM2__USB0_ID
+					>;
+					fsl,voltage = <MXS_VOLTAGE_HIGH>;
+					fsl,pull-up = <MXS_PULL_ENABLE>;
+				};
 			};
 
 			lcdif@80030000 {
@@ -131,7 +149,10 @@
 
 	ahb@80080000 {
 		usb0: usb@80080000 {
+			pinctrl-names = "default";
+			pinctrl-0 = <&apf28dev_usb0_id_pins>;
 			vbus-supply = <&reg_usb0_vbus>;
+			dr_mode = "otg";
 			status = "okay";
 		};
 
@@ -157,6 +178,7 @@
 			regulator-min-microvolt = <5000000>;
 			regulator-max-microvolt = <5000000>;
 			gpio = <&gpio1 23 1>;
+			enable-active-high;
 		};
 	};
 
@@ -177,4 +199,15 @@
 		brightness-levels = <0 4 8 16 32 64 128 255>;
 		default-brightness-level = <6>;
 	};
+
+	gpio-keys {
+		compatible = "gpio-keys";
+
+		user-button {
+			label = "User button";
+			gpios = <&gpio0 17 0>;
+			linux,code = <0x100>;
+			gpio-key,wakeup;
+		};
+	};
 };
