Fix SD slot numbering.

eMMC -> mmcblk2
microSD -> mmcblk1

Signed-off-by: Fabien Marteau <fabien.marteau@armadeus.com>

---
Index: linux-4.0-rc1/drivers/mmc/card/block.c
===================================================================
--- linux-4.0-rc1.orig/drivers/mmc/card/block.c
+++ linux-4.0-rc1/drivers/mmc/card/block.c
@@ -2161,7 +2161,7 @@ static struct mmc_blk_data *mmc_blk_allo
 	 */
 
 	snprintf(md->disk->disk_name, sizeof(md->disk->disk_name),
-		 "mmcblk%u%s", md->name_idx, subname ? subname : "");
+		 "mmcblk%u%s", card->host->index, subname ? subname : "");
 
 	if (mmc_card_mmc(card))
 		blk_queue_logical_block_size(md->queue.queue,
