Add allegro package to BR.

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: buildroot/package/Config.in
===================================================================
--- buildroot.orig/package/Config.in	2012-02-21 16:43:57.000000000 +0100
+++ buildroot/package/Config.in	2012-02-21 16:44:05.000000000 +0100
@@ -103,6 +103,7 @@
 source "package/rrdtool/Config.in"
 
 comment "graphic libraries"
+source "package/allegro/Config.in"
 source "package/directfb/Config.in"
 source "package/directfb-examples/Config.in"
 source "package/divine/Config.in"
Index: buildroot/package/allegro/Config.in
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/allegro/Config.in	2012-02-21 16:44:05.000000000 +0100
@@ -0,0 +1,13 @@
+config BR2_PACKAGE_ALLEGRO
+	bool "Allegro"
+	help
+	  Allegro is a game programming library for C/C++ developers
+	  distributed freely and supporting the following platforms:
+	  DOS, Unix (Linux, FreeBSD, Irix, Solaris, Darwin), Windows, QNX,
+	  BeOS and MacOS X.
+	  It provides many functions for graphics, sounds, player input
+	  (keyboard, mouse and joystick) and timers. It also provides fixed and
+	  floating point mathematical functions, 3d functions, file management
+	  functions, compressed datafile and a GUI.
+
+	  http://alleg.sourceforge.net/
Index: buildroot/package/allegro/allegro.mk
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/allegro/allegro.mk	2012-02-21 16:44:05.000000000 +0100
@@ -0,0 +1,75 @@
+#############################################################
+#
+# Allegro: a game programming library
+#
+#############################################################
+ALLEGRO_VERSION:=4.2.1
+ALLEGRO_SOURCE:=allegro-$(ALLEGRO_VERSION).tar.gz
+ALLEGRO_SITE:=http://downloads.sourceforge.net/alleg
+ALLEGRO_CAT:=zcat
+ALLEGRO_DIR:=$(BUILD_DIR)/allegro-$(ALLEGRO_VERSION)
+
+$(DL_DIR)/$(ALLEGRO_SOURCE):
+	$(call DOWNLOAD,$(ALLEGRO_SITE),$(ALLEGRO_SOURCE))
+
+allegro-source: $(DL_DIR)/$(ALLEGRO_SOURCE)
+
+$(ALLEGRO_DIR)/.unpacked: $(DL_DIR)/$(ALLEGRO_SOURCE)
+	$(ALLEGRO_CAT) $(DL_DIR)/$(ALLEGRO_SOURCE) | tar -C $(BUILD_DIR) $(TAR_OPTIONS) -
+	touch $@
+
+$(ALLEGRO_DIR)/.configured: $(ALLEGRO_DIR)/.unpacked
+	(cd $(ALLEGRO_DIR); \
+	$(TARGET_CONFIGURE_OPTS) \
+	CFLAGS="$(TARGET_CFLAGS)" \
+	LDFLAGS="$(TARGET_LDFLAGS)" \
+	CROSSCOMPILE=1 \
+	./configure \
+		--target=$(GNU_TARGET_NAME) \
+		--host=$(GNU_TARGET_NAME) \
+		--build=$(GNU_HOST_NAME) \
+		--prefix=/usr \
+		--without-svgalib \
+		--enable-modules=no \
+		--enable-static=yes \
+		--enable-shared=no \
+		--disable-asm \
+		--disable-mmx \
+		--disable-sse \
+		--disable-color24 \
+		--disable-color32 \
+		--without-x );
+	touch $@
+# allegro doesn't seem to build as shared library
+# 		--disable-color24 \
+# 		--disable-color32 \
+
+$(ALLEGRO_DIR)/lib/unix/liballeg.a: $(ALLEGRO_DIR)/.configured
+	$(MAKE) -C $(ALLEGRO_DIR) CROSSCOMPILE=1
+	touch $@
+
+#$(STAGING_DIR)/usr/lib/liballeg.so: $(ALLEGRO_DIR)/lib/unix/liballeg.a
+#	$(MAKE) -C $(ALLEGRO_DIR) install
+#	touch -c $@
+#
+# $(TARGET_DIR)/usr/lib/liballeg.so.so: $(STAGING_DIR)/usr/lib/liballeg.so
+# 	cp -dpf $(STAGING_DIR)/usr/lib/liballeg*.so* $(TARGET_DIR)/usr/lib/
+# 	-$(STRIP) --strip-unneeded $(TARGET_DIR)/usr/lib/liballeg.so
+
+ALLEGRO allegro: $(ALLEGRO_DIR)/lib/unix/liballeg.a
+
+allegro-clean:
+	$(MAKE) DESTDIR=$(TARGET_DIR) CC=$(TARGET_CC) -C $(ALLEGRO_DIR) uninstall
+	-$(MAKE) -C $(ALLEGRO_DIR) clean
+
+allegro-dirclean:
+	rm -rf $(ALLEGRO_DIR)
+
+#############################################################
+#
+# Toplevel Makefile options
+#
+#############################################################
+ifeq ($(strip $(BR2_PACKAGE_ALLEGRO)),y)
+TARGETS+=allegro
+endif
