Make uClibc 0.9.33 as Buildroot's default version

Signed-off-by: Jérémie Scheer <jeremie.scheer@armadeus.com>

Index: buildroot/toolchain/uClibc/uclibc.mk
===================================================================
--- buildroot.orig/toolchain/uClibc/uclibc.mk	2012-02-29 23:19:47.000000000 +0100
+++ buildroot/toolchain/uClibc/uclibc.mk	2012-03-23 16:57:54.000000000 +0100
@@ -102,23 +102,6 @@
 		-e 's,^SHARED_LIB_LOADER_PREFIX=.*,SHARED_LIB_LOADER_PREFIX=\"/lib\",g' \
 		$(UCLIBC_DIR)/.oldconfig
 ifeq ($(UCLIBC_TARGET_ARCH),arm)
-	(/bin/echo "# CONFIG_GENERIC_ARM is not set"; \
-	 /bin/echo "# CONFIG_ARM610 is not set"; \
-	 /bin/echo "# CONFIG_ARM710 is not set"; \
-	 /bin/echo "# CONFIG_ARM7TDMI is not set"; \
-	 /bin/echo "# CONFIG_ARM720T is not set"; \
-	 /bin/echo "# CONFIG_ARM920T is not set"; \
-	 /bin/echo "# CONFIG_ARM922T is not set"; \
-	 /bin/echo "# CONFIG_ARM926T is not set"; \
-	 /bin/echo "# CONFIG_ARM10T is not set"; \
-	 /bin/echo "# CONFIG_ARM1136JF_S is not set"; \
-	 /bin/echo "# CONFIG_ARM1176JZ_S is not set"; \
-	 /bin/echo "# CONFIG_ARM1176JZF_S is not set"; \
-	 /bin/echo "# CONFIG_ARM_SA110 is not set"; \
-	 /bin/echo "# CONFIG_ARM_SA1100 is not set"; \
-	 /bin/echo "# CONFIG_ARM_XSCALE is not set"; \
-	 /bin/echo "# CONFIG_ARM_IWMMXT is not set"; \
-	) >> $(UCLIBC_DIR)/.oldconfig
 	$(SED) 's/^\(CONFIG_[^_]*[_]*ARM[^=]*\)=.*/# \1 is not set/g' \
 		 $(UCLIBC_DIR)/.oldconfig
 	$(SED) 's/^.*$(UCLIBC_ARM_TYPE).*/$(UCLIBC_ARM_TYPE)=y/g' $(UCLIBC_DIR)/.oldconfig
Index: buildroot/toolchain/uClibc/Config.in
===================================================================
--- buildroot.orig/toolchain/uClibc/Config.in	2012-03-23 17:04:18.000000000 +0100
+++ buildroot/toolchain/uClibc/Config.in	2012-03-23 17:04:28.000000000 +0100
@@ -6,7 +6,7 @@
 choice
 	prompt "uClibc C library Version"
 	default BR2_UCLIBC_VERSION_0_9_31 if (BR2_avr32 || BR2_sh) # 0.9.32 broken
-	default BR2_UCLIBC_VERSION_0_9_32
+	default BR2_UCLIBC_VERSION_0_9_33
 	help
 	  Select the version of uClibc you wish to use.
 
