Add Flite package (speech synthesis engine).

Signed-off-by: Julien Boibessot <julien.boibessot@armadeus.com>

Index: buildroot/package/multimedia/Config.in
===================================================================
--- buildroot.orig/package/multimedia/Config.in	2012-02-21 16:43:27.000000000 +0100
+++ buildroot/package/multimedia/Config.in	2012-02-21 16:44:45.000000000 +0100
@@ -4,6 +4,7 @@
 source "package/multimedia/aumix/Config.in"
 source "package/multimedia/faad2/Config.in"
 source "package/multimedia/flac/Config.in"
+source "package/multimedia/flite/Config.in"
 source "package/multimedia/ffmpeg/Config.in"
 source "package/multimedia/gstreamer/Config.in"
 source "package/multimedia/gst-ffmpeg/Config.in"
Index: buildroot/package/multimedia/flite/Config.in
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/multimedia/flite/Config.in	2012-02-21 16:44:45.000000000 +0100
@@ -0,0 +1,11 @@
+config BR2_PACKAGE_FLITE
+	bool "flite"
+	help
+	  Flite: a small, fast run time synthesis engine.
+	  Flite (festival-lite) is a small, fast run-time synthesis engine
+	  developed at CMU and primarily designed for small embedded machines
+	  and/or large servers. Flite is designed as an alternative synthesis
+	  engine to Festival for voices built using the FestVox suite of voice
+	  building tools.
+
+	  http://www.speech.cs.cmu.edu/flite
Index: buildroot/package/multimedia/flite/flite.mk
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/multimedia/flite/flite.mk	2012-02-21 16:44:45.000000000 +0100
@@ -0,0 +1,44 @@
+#############################################################
+#
+# flite
+#
+#############################################################
+
+FLITE_VERSION = 1.4
+FLITE_SOURCE = flite-$(FLITE_VERSION)-release.tar.bz2
+FLITE_SITE = http://www.speech.cs.cmu.edu/flite/packed/flite-$(FLITE_VERSION)/
+FLITE_INSTALL_STAGING = YES
+FLITE_INSTALL_TARGET = YES
+
+FLITE_CONF_OPT = --enable-shared
+FLITE_INSTALL_STAGING_OPT = INSTALLBINDIR=$(STAGING_DIR)/usr/bin INSTALLLIBDIR=$(STAGING_DIR)/usr/lib INSTALLINCDIR=$(STAGING_DIR)/usr/include/flite install
+FLITE_INSTALL_TARGET_OPT = INSTALLBINDIR=$(TARGET_DIR)/usr/bin INSTALLLIBDIR=$(TARGET_DIR)/usr/lib INSTALLINCDIR=$(STAGING_DIR)/usr/include/flite install
+
+$(eval $(call AUTOTARGETS))
+
+# Flite (strange) archive format makes Buildroot failed to extract it (--strip-components).
+# So do it manually:
+$(FLITE_TARGET_EXTRACT): $(FLITE_TARGET_SOURCE)
+	$(call MESSAGE,"Extracting")
+	$(INFLATE$(suffix $($(PKG)_SOURCE))) $(DL_DIR)/$($(PKG)_SOURCE) | \
+		$(TAR) -C $(BUILD_DIR) $(TAR_OPTIONS) -
+	$(Q)mv $(@D)/.stamp* $(BUILD_DIR)/flite-$(FLITE_VERSION)-release/
+	$(Q)rm -rf $(@D)
+	$(Q)mv $(BUILD_DIR)/flite-$(FLITE_VERSION)-release $(@D)
+# some packages have messed up permissions inside
+	$(Q)chmod -R ug+rw $(@D)
+	touch $@
+
+$(FLITE_HOOK_POST_INSTALL): $(FLITE_TARGET_INSTALL_TARGET)
+	rm -f $(TARGET_DIR)/usr/lib/libflite*.a
+ifeq ($(BR2_ENABLE_DEBUG),)
+	$(STRIPCMD) $(STRIP_STRIP_UNNEEDED) $(TARGET_DIR)/usr/lib/libflite*.so.$(FLITE_VERSION)
+	$(STRIPCMD) $(STRIP_STRIP_UNNEEDED) $(TARGET_DIR)/usr/bin/flite*
+endif
+	touch $@
+
+# Flite has no uninstall target
+$(FLITE_TARGET_UNINSTALL):
+	$(call MESSAGE,"Uninstalling")
+	rm -f $(TARGET_DIR)/usr/lib/libflite*
+	rm -f $(TARGET_DIR)/usr/bin/flite*
