Add CUnit package.

Signed-off-by: Jérémie Scheer <jeremie.scheer@armadeus.com>

===================================================================
Index: buildroot/package/cunit/Config.in
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/cunit/Config.in	2012-10-11 11:43:27.000000000 +0200
@@ -0,0 +1,4 @@
+config BR2_PACKAGE_CUNIT
+	bool "CUnit library"
+	help
+	  Compile & install CUnit library.
Index: buildroot/package/cunit/cunit.mk
===================================================================
--- /dev/null	1970-01-01 00:00:00.000000000 +0000
+++ buildroot/package/cunit/cunit.mk	2012-10-11 14:16:48.000000000 +0200
@@ -0,0 +1,18 @@
+#############################################################
+#
+# CUnit
+#
+#############################################################
+
+CUNIT_VERSION = 156
+CUNIT_SITE = https://cunit.svn.sourceforge.net/svnroot/cunit
+CUNIT_SITE_METHOD = svn
+CUNIT_INSTALL_STAGING = YES
+
+define CUNIT_BOOTSTRAP_CMD
+	(cd $(@D); ./bootstrap)
+endef
+
+CUNIT_PRE_CONFIGURE_HOOKS += CUNIT_BOOTSTRAP_CMD
+
+$(eval $(call GENTARGETS,package,cunit))
Index: buildroot/package/Config.in
===================================================================
--- buildroot.orig/package/Config.in	2012-10-11 11:23:28.000000000 +0200
+++ buildroot/package/Config.in	2012-10-11 14:20:32.000000000 +0200
@@ -351,6 +351,7 @@
 source "package/libglib2/Config.in"
 source "package/liboil/Config.in"
 source "package/startup-notification/Config.in"
+source "package/cunit/Config.in"
 endmenu
 
 menu "Text and terminal handling"
