Index: linux-3.8-rc1/arch/arm/mach-imx/iomux-mx1.h
===================================================================
--- linux-3.8-rc1.orig/arch/arm/mach-imx/iomux-mx1.h	2012-12-22 02:19:00.000000000 +0100
+++ linux-3.8-rc1/arch/arm/mach-imx/iomux-mx1.h	2013-01-01 21:19:56.000000000 +0100
@@ -119,7 +119,7 @@
 #define PD6_PF_LSCLK		(GPIO_PORTD | GPIO_PF | GPIO_OUT | 6)
 #define PD7_PF_REV		(GPIO_PORTD | GPIO_PF | 7)
 #define PD7_AF_UART2_DTR	(GPIO_PORTD | GPIO_AF | GPIO_IN | 7)
-#define PD7_AIN_SPI2_SCLK	(GPIO_PORTD | GPIO_AIN | 7)
+#define PD7_AIN_SPI2_SCLK	(GPIO_PORTD | GPIO_AIN | GPIO_OUT | 7)
 #define PD8_PF_CLS		(GPIO_PORTD | GPIO_PF | 8)
 #define PD8_AF_UART2_DCD	(GPIO_PORTD | GPIO_AF | GPIO_OUT | 8)
 #define PD8_AIN_SPI2_SS		(GPIO_PORTD | GPIO_AIN | 8)
