
Signed-off-by: Gwenhael Goavec-Merou <gwenhael.goavec-merou@armadeus.com>
---
Index: linux-3.12-rc1/arch/arm/boot/dts/imx27-apf27dev.dts
===================================================================
--- linux-3.12-rc1.orig/arch/arm/boot/dts/imx27-apf27dev.dts
+++ linux-3.12-rc1/arch/arm/boot/dts/imx27-apf27dev.dts
@@ -108,7 +108,16 @@
 				MX27_PAD_CSPI2_SS0__GPIO4_21 0
 				MX27_PAD_CSPI2_SS1__GPIO4_20 0
 				MX27_PAD_CSI_D5__GPIO2_17 0
+				MX27_PAD_TOUT__GPIO3_14 0
 			>;
 		};
 	};
 };
+
+&sdhci2 {
+	bus-width = <4>;
+	cd-gpios = <&gpio3 14 0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sdhc2_1>;
+	status = "okay";
+};
